### PR TITLE
feat: display token usage in status bar and webview

### PR DESCRIPTION
**Pull request by terrygzhou**
## Summary

Adds token usage visualization so users can see context consumption in real-time.

## Changes

**3 files changed, +151/-3**

### `SessionManager.ts`
- Import `UsageUpdate` type from ACP SDK v0.21.1
- Add `usage: UsageUpdate | null` to `SessionInfo` interface
- `pendingUsages` buffer map for notifications that arrive before session registration
- `applyUsage()` method with pending→drain lifecycle (mirrors `pendingTitles` pattern)
- `usage: null` in all 3 session creation paths (create, load, resume)

### `StatusBarManager.ts`
- Subscribe to `session-usage-changed` event
- Append token display: `\$(hubot) ACP: AgentName  45K/200K`
- `formatTokens()` helper with K/M suffix

### `ChatWebviewProvider.ts`
- Handle `usage_update` session notifications via `applyUsage()`
- Send `usage` in `sendCurrentState()` for initial webview hydration
- Input area progress bar: brain icon + fill track + token count text
- `updateUsageBar()` + `formatTokens()` webview JS helpers
- Restore usage bar from state on reconnect; hide on disconnect/clear

## Data Flow

```
Agent → session/update { sessionUpdate: "usage_update", used, size }
  → AcpClientImpl.sessionUpdate()
    → SessionUpdateHandler.handleUpdate()
      → ChatWebviewProvider.handleSessionUpdate()
        → SessionManager.applyUsage()         ← stores session.usage
        → postMessage({ type: 'sessionUpdate' }) → webview updateUsageBar()
      → SessionManager.emit('session-usage-changed') → StatusBarManager
```

## UI

**Status bar:** `\$(hubot) ACP: Qwen3.6-27B  45K/200K`

**Webview (input area):**
```
[mode picker] [model picker] [...]
[🧠 ████░░░░░░░░░  45K/200K ]
[ Type a message...              ]
```

## Notes
- Graceful: if the agent doesn't emit `usage_update`, status bar and webview show nothing extra (no regression)
- Verified with SGLang/Qwen3.6-27B agent

### CODE REVIEW

File: src/ui/ChatWebviewProvider.ts [hunk 1]
```diff
@@ -2199,6 +2297,12 @@ export class ChatWebviewProvider implements vscode.WebviewViewProvider {
         availableCommands = session.availableCommands;
       }
       updatePlaceholder();
+
+      // Restore usage bar
+      if (session.usage) {
+        lastUsage = session.usage;
+        updateUsageBar(lastUsage);
+      }
```
Comment (formulahendry) [inline src/ui/ChatWebviewProvider.ts]: The usage bar keeps showing the previous session's numbers when switching sessions.

`showSessionConnected()` is reached via `notifyActiveSessionChanged()` → `sendCurrentState()`. When switching from a session that has usage to one that hasn't received a `usage_update` yet, `session.usage` is `null`, so this whole block is skipped and the bar keeps rendering the **previous** session's fill width and token count — the user would think the new session already consumed 45K of context.

Note that `setConfigOptionsState([])` a few lines above explicitly handles the empty case, so resetting on switch appears to be the intended pattern here; `usage` just missed it.

`updateUsageBar()` already handles falsy input (it calls `classList.remove('visible')`), so dropping the guard is enough:

```js
// Restore usage bar
lastUsage = session.usage || null;
updateUsageBar(lastUsage);
```

File: src/ui/ChatWebviewProvider.ts [hunk 2]
```diff
@@ -419,6 +423,38 @@ export class ChatWebviewProvider implements vscode.WebviewViewProvider {
...
+      height: 22px;
+      font-size: 0.8em;
+      color: var(--vscode-descriptionForeground);
+    }
+    .input-usage-bar.visible { display: flex; }
+    .input-usage-label { flex-shrink: 0; }
+    .input-usage-track {
+      flex: 1;
+      height: 6px;
+      background: var(--vscode-input-border, color-mix(in srgb, var(--vscode-input-border) 30%, transparent));
```
Comment (formulahendry) [inline src/ui/ChatWebviewProvider.ts]: This `var()` fallback references the same variable it is falling back from, so neither branch produces the intended 30% tint.

The fallback is only used when `--vscode-input-border` is **undefined** — but the fallback itself calls `var(--vscode-input-border)` with no default, which is then invalid at computed-value time and drops the whole `background` declaration.

I checked the built-in themes in `microsoft/vscode`:

| Theme | `input.border` | Result |
|---|---|---|
| Dark Modern / Light Modern (current defaults) | defined | Renders, but as an opaque `#3C3C3C`, not the intended 30% tint |
| Dark+ / Light+ / Dark (Visual Studio) / Light (Visual Studio) | **not defined** | `background` is dropped → track is fully transparent |

Most third-party themes don't define `input.border` either, so the track will often be invisible (the fill still renders, so it looks like a floating bar).

Putting the fallback inside `color-mix()` fixes both paths:

```css
background: color-mix(in srgb, var(--vscode-input-border, #808080) 30%, transparent);
```

The same declaration is duplicated at line 829, so it needs the same fix there (or the duplicate block can just be removed).

**Review by formulahendry — COMMENTED** [review]
See inline comments.

File: HANDOFF.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# Session Handoff — 2026-07-27 11:30 UTC
+
+**Project**: `vscode-acp` | `/home/terry/workspace/projects/vscode-acp`
+**Branch**: `feature/usage-display`
+**Last commit**: `9fba260 auto: handoff`
+**Uncommitted**: 6 files: `M package-lock.json`, `M package.json`, `M src/ui/ChatWebviewProvider.ts`, `M src/ui/StatusBarManager.ts`, `?? test/`
```
Comment (formulahendry) [inline HANDOFF.md]: This file looks like it was committed by accident — it comes from the four `auto: handoff` commits and is a local agent-session artifact rather than part of this feature.

A few reasons it shouldn't land on `main`:

- It hardcodes your local absolute path (`/home/terry/workspace/projects/vscode-acp`).
- The "Recent Goals / Recent Actions" content is unrelated to token usage display (it's about introducing ACP and MCP).
- It references an internal session file name (`20260616_132003_5c76a01a.jsonl`) that means nothing to this repo.
- It's a snapshot of transient state, so it will go stale immediately and generate noisy diffs on every future handoff.

Could you drop it from this PR and add it (or whatever directory your tool writes it to) to `.gitignore`? It's currently not ignored, so it'll keep getting picked up.

Unrelated but worth asking since this file surfaced it: the uncommitted list mentions `?? test/`. If that's the test coverage for this feature, it'd be great to include it here.

File: src/ui/ChatWebviewProvider.ts [hunk 4]
```diff
@@ -1269,6 +1367,10 @@ export class ChatWebviewProvider implements vscode.WebviewViewProvider {
       chatHistory = saved.chatHistory || [];
       sessionState = saved.sessionState || null;
       hasActiveSession = saved.hasActiveSession || false;
+      if (saved.lastUsage) {
+        lastUsage = saved.lastUsage;
+        updateUsageBar(lastUsage);
+      }
```
Comment (formulahendry) [inline src/ui/ChatWebviewProvider.ts]: **Option A of two — see my comment on `saveState()` (line 1360) for Option B; only one of them is needed.**

This restore runs before the `hasActiveSession` check below, so the usage bar is shown unconditionally. When the webview is restored while no agent is connected, `hasActiveSession` is `false` and the session banner stays hidden, but the bar still renders the token counts from the previous session.

It's the same class of staleness as the session-switch bug fixed in this commit, just entering through a different path. Two things make it more visible than it first appears:

1. The webview is created with `enableScripts: true` only — no `retainContextWhenHidden` — so it is disposed whenever the view is hidden. `restoreState()` therefore runs every time the user switches back to the ACP panel, not just on window reload.
2. The `.input-area.disabled` rules only dim `.input-toolbar`, `.input-editor-wrap` and `.input-send-row`. `#usageBar` is a sibling of those inside `.input-area`, so it renders at full opacity while everything around it is greyed out.

The extension side does self-correct — `sendCurrentState()` posts `{ session: null }`, which calls `showNoSession()` and hides the bar — so this is a flash rather than a persistent wrong state. But it's a flash that repeats on every panel switch.

Moving the restore inside the existing branch keeps it consistent with how the rest of the state is restored:

```js
if (hasActiveSession && sessionState) {
  showSessionConnectedFromState(sessionState);
  if (saved.lastUsage) {
    lastUsage = saved.lastUsage;
    updateUsageBar(lastUsage);
  }
}
```

Worth adding `.input-area.disabled .input-usage-bar` to the dim rule regardless, so the bar can't sit at full brightness in a disabled input area.

File: src/ui/ChatWebviewProvider.ts [hunk 5]
```diff
@@ -1254,12 +1333,31 @@ export class ChatWebviewProvider implements vscode.WebviewViewProvider {
...
+      usageBarText.textContent = formatTokens(usage.used) + '/' + formatTokens(usage.size);
+    }
+
     // --- State persistence ---
     let chatHistory = [];
     let sessionState = null;
 
     function saveState() {
-      vscode.setState({ chatHistory, sessionState, hasActiveSession });
+      vscode.setState({ chatHistory, sessionState, hasActiveSession, lastUsage });
```
Comment (formulahendry) [inline src/ui/ChatWebviewProvider.ts]: **Option B of two — the alternative to my comment on `restoreState()` (lines 1370-1373). Pick whichever you prefer; both fix the same problem.**

Stepping back: is persisting `lastUsage` worth keeping at all?

`sendCurrentState()` now sends `session.usage` to the webview, so after any webview rebuild the real value arrives on its own. Persistence only saves the few tens of milliseconds before that message lands, and in exchange it adds a second source of truth that has to stay in sync with the first.

It isn't quite in sync today: the `usage_update` handler assigns `lastUsage` but doesn't call `saveState()`, so what actually gets written is whatever value happened to be current the last time some other code path triggered a save. Fixing that by calling `saveState()` on every `usage_update` isn't free either — it re-serialises the whole `chatHistory`, which can be large and updates frequently during streaming. Throttling it would work, but that's more machinery for a very small benefit.

Dropping `lastUsage` from `setState()`/`getState()` removes the stale-bar issue and the sync problem in one go, and the bar still populates correctly from `sendCurrentState()`. That said, if you'd rather keep the persistence for the smoother restore, the fix in the `restoreState()` comment is entirely sufficient — no need to do both.

**Review by formulahendry — COMMENTED** [review]
See inline comments. Thanks for the quick turnaround — both items from my previous review are cleanly addressed, and you fixed the status bar `size === 0` case too.

**Review by formulahendry — COMMENTED** [review]
See inline comments.

File: test/usage.test.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
+import { describe, it } from 'mocha';
+import { strictEqual } from 'assert';
+
+/** Replicate StatusBarManager.formatTokens for unit testing. */
```
Comment (formulahendry) [inline test/usage.test.ts]: These tests don't exercise the production code — they re-implement it. `formatTokens`, `computeUsageSuffix`, and `simulateRestoreState` are all hand-copied duplicates of logic that lives in `StatusBarManager.ts` and `ChatWebviewProvider.ts`. A regression in either of those files would leave all 143 lines of this suite green, so the suite currently only protects itself.

Minimal fix: export the real helpers and import them here.

```ts
// StatusBarManager.ts
export function formatTokens(n: number): string { ... }
export function computeUsageSuffix(usage?: UsageUpdate | null): string { ... }
```

```ts
// test/usage.test.ts
import { formatTokens, computeUsageSuffix } from '../src/ui/StatusBarManager';
```

For the webview side, the same exported `formatTokens` can be injected into the HTML template instead of being written out a second time — which also removes the divergence risk that the `formatTokens (webview)` block is trying to guard against.

File: package.json [hunk 7]
```diff
@@ -417,6 +418,7 @@
     "@vscode/test-cli": "^0.0.10",
     "@vscode/test-electron": "^2.4.1",
     "@vscode/vsce": "^3.7.1",
+    "chai": "^6.2.2",
```
Comment (formulahendry) [inline package.json]: `chai` and `@types/chai` are added here but never used. `test/usage.test.ts` asserts with Node's built-in `assert` (`strictEqual`), and a repo-wide search for `chai` returns 0 hits outside `package.json` / `package-lock.json`.

Either drop both devDependencies, or switch the assertions over to chai — but let's not carry both.

File: package.json [hunk 8]
```diff
@@ -397,7 +397,7 @@
     "compile": "webpack",
     "watch": "webpack --watch",
     "package": "webpack --mode production --devtool hidden-source-map",
-    "compile-tests": "tsc -p . --outDir out",
+    "compile-tests": "tsc -p tsconfig.test.json --outDir out",
```
Comment (formulahendry) [inline package.json:400]: 🔴 **Blocker — this silently excludes the repo's existing integration tests.**

`tsconfig.json` has `rootDir: "src"`; `tsconfig.test.json` has `rootDir: "."`. With `--outDir out`, that relocates every emitted file:

| config | emit for `src/test/extension.test.ts` | matches `out/test/**/*.test.js` in `.vscode-test.mjs`? |
|---|---|---|
| `tsc -p .` (before) | `out/test/extension.test.js` | ✅ |
| `tsc -p tsconfig.test.json` (after) | `out/src/test/extension.test.js` | ❌ |

I verified this by compiling both ways locally. What stops running is the repo's only integration test file — extension presence, successful activation, and registration of `acp.connectAgent` / `acp.newConversation` / `acp.openChat`. Three tests just quietly disappear from the run; nothing fails, so CI would stay green.

Two ways out:
1. Widen the glob in `.vscode-test.mjs` to `out/**/*.test.js`, or
2. Move `test/usage.test.ts` under `src/test/` and keep `tsc -p .`.

Option 1 is the smaller diff, but please double-check it doesn't pick up anything unintended.

File: src/ui/ChatWebviewProvider.ts [hunk 9]
```diff
@@ -1274,6 +1372,13 @@ export class ChatWebviewProvider implements vscode.WebviewViewProvider {
         showSessionConnectedFromState(sessionState);
       }
 
+      // Only restore usage bar when there is an active session —
+      // otherwise we show stale counts from a previous disconnected session.
+      if (hasActiveSession && sessionState?.usage) {
```
Comment (formulahendry) [inline src/ui/ChatWebviewProvider.ts]: 🟠 **Major — dead code: this branch can never execute.**

`sessionState` is written in exactly one place in this file (the `saveState()` object literal, ~L2283-2287), and it carries only `agentName`, `cwd`, and `title`. `usage` is never assigned to it — a search for `usage` within the persisted shape returns 0 hits — so `sessionState?.usage` is always `undefined` and this `if` body is unreachable.

The bar still displays correctly, because the real path is `sendCurrentState()` → `showSessionConnected()`, which reads `session.usage` (live session object, not persisted state).

So either:
- **Drop this block** (and the mirroring test in `test/usage.test.ts`), since `showSessionConnected()` already covers it; or
- **Add `usage` to the `saveState()` payload** if the intent really is to restore the bar before the session reconnects.

Worth deciding explicitly — right now the code reads as if persistence works, but it doesn't.

File: test/usage.test.ts [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+function simulateRestoreState(
+  hasActiveSession: boolean,
+  sessionState: { usage?: { used: number; size: number } } | null,
+): RestoreStateResult {
+  const usage = (hasActiveSession && sessionState?.usage)
+    ? sessionState.usage
+    : null;
+  return { barRestored: usage !== null, usage };
+}
```
Comment (formulahendry) [inline test/usage.test.ts]: 🟠 **Major — this mirror test conceals the dead code above rather than covering it.**

Two problems compound here:

1. **It's a hand-copied duplicate.** `simulateRestoreState` (L20-28) and `formatTokensWebview` (L7-11) reimplement the webview logic in the test file. If someone breaks the real code inside the inline webview HTML, these tests stay green — they only test themselves. The comment says "can't import directly", which is fair, but a mirror that can silently diverge is arguably worse than no test.

2. **The first case tests an impossible scenario.** L98-102 passes `{ usage: { used, size } }` and asserts `barRestored === true`, but as noted on `ChatWebviewProvider.ts:1377`, `sessionState` never carries `usage` in production. So this asserts behaviour that cannot occur.

The other four cases (`false`/`{}`/`null` → not restored) all pass trivially for the same reason.

If the dead branch gets removed, this whole `describe` block should go with it. If `usage` gets persisted instead, this test becomes meaningful — but consider extracting the gating logic into `src/util/usage.ts` so it can be imported rather than mirrored.

File: package.json [hunk 11]
```diff
@@ -397,7 +397,7 @@
     "compile": "webpack",
     "watch": "webpack --watch",
     "package": "webpack --mode production --devtool hidden-source-map",
-    "compile-tests": "tsc -p . --outDir out",
+    "compile-tests": "tsc -p tsconfig.test.json --outDir out",
     "watch-tests": "tsc -p . -w --outDir out",
```
Comment (formulahendry) [inline package.json]: 🟡 **Minor — `watch-tests` is now inconsistent with `compile-tests`.**

`compile-tests` moved to `tsconfig.test.json` while this one still uses `tsc -p .`, so watch mode and one-shot compilation emit to different layouts. Whichever resolution is chosen for the blocker on L400, this line should follow it so `npm run watch-tests` reproduces what `pretest` builds.

**Review by formulahendry — COMMENTED** [review]
## Summary

Third pass. The four new commits addressed all three comments from the previous round — `HANDOFF.md` is gone and now gitignored, the `test/` directory is committed, and the persistence path was simplified per Option B. Thanks for the quick turnaround.

The usage-bar feature itself looks correct, and I ran the new suite locally: **27 passing**. But this round surfaced one blocker plus two related issues that are best fixed together.

**🔴 The test-infrastructure change silently drops the existing integration tests.** Switching `compile-tests` to `tsconfig.test.json` changes `rootDir` from `src` to `.`, so `src/test/extension.test.ts` now emits to `out/src/test/extension.test.js` instead of `out/test/extension.test.js` — which no longer matches the `out/test/**/*.test.js` glob in `.vscode-test.mjs`. The repo's only integration tests (extension presence, activation, and command registration) stop running, and nothing turns red while it happens. Details inline on `package.json`.

**🟠 The `restoreState()` usage branch is dead code, and the new test conceals it.** `sessionState` is built in exactly one place in the file and carries only `agentName` / `cwd` / `title`; `usage` is never written into it, so `if (hasActiveSession && sessionState?.usage)` can never fire. Display is unaffected — `sendCurrentState()` → `showSessionConnected()` fills the bar from `session.usage`. The real problem is that `simulateRestoreState()` in the test is a hand-copied mirror asserting a scenario that cannot occur in production code, so it stays green no matter what the webview does. Either drop the dead branch together with its mirror test, or actually persist `usage` into `sessionState` if restore-before-connect was the intent.

The rest is minor, though one is worth calling out: I exhaustively compared both `formatTokens` implementations across `1_000..999_999` and found **zero** differing inputs — so the test named *"Math.round differs from toFixed"* asserts a distinction that does not exist.

**Meta:** CI reports 0 check runs on both `d9a0aa55` and `3f93183d`, whereas the earlier `9825d446` had all three platforms green. The triggers in `ci.yml` look correct, so this may be a fork PR awaiting manual workflow approval — worth checking. (Note that CI would not have caught the blocker anyway: the tests are excluded, not failing.)

File: test/usage.test.ts [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        strictEqual(formatTokens(input), expected);
+      });
+    }
+  });
+
+  describe('formatTokens (webview)', () => {
+    it('rounds 1500 to "2K"', () => {
+      strictEqual(formatTokensWebview(1_500), '2K');
+    });
+    it('rounds 1499 to "1K" (Math.round differs from toFixed)', () => {
```
Comment (formulahendry) [inline test/usage.test.ts]: 🟡 **Minor — this test name asserts a difference that doesn't exist.**

I brute-forced every integer in `1_000..999_999` through both implementations:

- `src/util/usage.ts`: `(n / 1_000).toFixed(0) + 'K'`
- webview: `Math.round(n / 1_000) + 'K'`

**Zero** differing inputs. `toFixed(0)` and `Math.round` do diverge in general (banker's-rounding-adjacent edge cases, e.g. `(1.005).toFixed(2)`), but not over this domain for this expression.

The assertion itself (`1_499` → `'1K'`) is correct, so the test passes — it's the name that's misleading. Suggest renaming to something like `'rounds 1499 down to "1K"'`, since a future reader will otherwise go looking for a divergence that isn't there.

File: src/util/usage.ts [hunk 13]
```diff
@@ -0,0 +1,18 @@
+/** Token usage helpers — pure functions, no VS Code dependencies. */
+
+/** Format token count with K/M suffix. */
+export function formatTokens(n: number): string {
+  if (n >= 1_000_000) { return (n / 1_000_000).toFixed(1) + 'M'; }
+  if (n >= 1_000) { return (n / 1_000).toFixed(0) + 'K'; }
+  return String(n);
+}
```
Comment (formulahendry) [inline src/util/usage.ts:8]: 🟡 **Minor — `formatTokens` now exists in three places.**

1. Here (status bar, `toFixed(0)`)
2. `ChatWebviewProvider.ts:1337-1341` (webview, `Math.round`)
3. `test/usage.test.ts:7-11` (test mirror of #2)

Copy #2 is genuinely hard to avoid — it lives in inline webview HTML with no module loader. But copy #3 is avoidable, and the three can drift independently.

One option: keep this file as the single source and inject it into the webview HTML via template interpolation (e.g. `${formatTokens.toString()}`) so the webview and the tests both derive from one definition. Not a blocker, just flagging that the count went from 1 to 3 in this PR.

File: test/usage.test.ts [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+  describe('formatTokens (status bar)', () => {
+    const cases: [number, string][] = [
+      [0, '0'],
+      [42, '42'],
+      [999, '999'],
+      [1_000, '1K'],
+      [1_500, '2K'],
+      [45_000, '45K'],
+      [199_999, '200K'],
+      [999_999, '1000K'],
```
Comment (formulahendry) [inline test/usage.test.ts:24]: 🔵 **Nit — this case locks in a display glitch as the expected value.**

`999_999` → `'1000K'` is what the code does today, but `1000K` is not something you'd want a user to see; `1.0M` would be the natural rendering.

Encoding it as an expectation makes it harder to fix later — whoever improves the formatting has to change a test that looks deliberate. If the current output is intentional, a short comment saying so would help. Otherwise consider dropping this case or changing the threshold to `>= 999_500`.

File: src/ui/ChatWebviewProvider.ts [hunk 15]
```diff
@@ -773,7 +809,39 @@ export class ChatWebviewProvider implements vscode.WebviewViewProvider {
...
+      background: var(--vscode-progressBar-background);
+      border-radius: 3px;
+      transition: width 0.3s ease;
+    }
+    .input-usage-text {
+      flex-shrink: 0;
+      min-width: 70px;
+      text-align: right;
+      font-variant-numeric: tabular-nums;
+    }
```
Comment (formulahendry) [inline src/ui/ChatWebviewProvider.ts:844]: 💡 **Suggestion — the duplicated CSS block from the previous round is still here.**

This is a verbatim copy of the rules added at L405 and L426-456; `.input-usage-track` alone now appears three times in the file. I raised this last round and understand there are two separate HTML templates involved, so it may be intentional.

If so, no action needed. If not, extracting the shared rules into a single string constant that both templates interpolate would keep them from drifting — the `color-mix(...)` fallback and the `22px` height are exactly the kind of values that get updated in one copy and not the other.

**Review by formulahendry — COMMENTED** [review]
See inline comments.

File: .vscode/launch.json [hunk 16]
```diff
@@ -23,10 +23,10 @@
 			"request": "launch",
 			"args": [
 				"--extensionDevelopmentPath=${workspaceFolder}",
-				"--extensionTestsPath=${workspaceFolder}/out/test"
+				"--extensionTestsPath=${workspaceFolder}/out"
```
Comment (formulahendry) [inline .vscode/launch.json:26]: This path is not a runnable extension-test entry point. `--extensionTestsPath` must resolve to a module that exports `run()`, but the compiled output contains only the individual test files and no `out/index.js`; resolving `out/` instead follows `out/package.json` to the nonexistent `out/dist/extension.js`, so launching **Extension Tests** from VS Code fails with `MODULE_NOT_FOUND`. Please add a standard compiled test runner and point this argument to it (or switch the launch configuration to the supported `@vscode/test-cli` debug setup).
